### PR TITLE
Allow boundary be associated with child elements 

### PR DESCRIPTION
**Pull request by dewenyushu**
Taking over #3094

### CODE REVIEW

**Comment by dewenyushu** [comment]
@roystgnr I just tried to address your concerns for the original PR (#3094 ) here. Please let me know your comments. Appreciate it so much!

File: src/mesh/boundary_info.C [hunk 1]
```diff
@@ -1712,6 +1757,47 @@ BoundaryInfo::build_shellface_boundary_ids(std::vector<boundary_id_type> & b_ids
...
+            this->add_side(parent, side, bnd_id);
+            return;
+          }
+      }
+    }
+    // No relative share the same boundary, therefore, we remove it from the parent, if any.
+    this->remove_side(parent, side, bnd_id);
+  }
+}
+
```
Comment (dewenyushu) [inline src/mesh/boundary_info.C:1800]: Here is the function to transfer boundary to the parent before coarsening

File: src/mesh/mesh_refinement.C [hunk 2]
```diff
@@ -1371,6 +1371,11 @@ bool MeshRefinement::_coarsen_elements ()
 
   for (auto & elem : _mesh.element_ptr_range())
     {
+      // Make sure we transfer the element's boundary id(s)
+      // up to its parent when necessary before coarsening.
+      // This can be adding or removing the corresonding boundary info.
+      _mesh.get_boundary_info().transfer_boundary_ids_to_parent(elem);
```
Comment (dewenyushu) [inline src/mesh/mesh_refinement.C:1377]: Here is when I transfer the boundary info to the parent

**Comment by dewenyushu** [comment]
I tried numerous times adding a unit test for this code change and it would not work.. Not sure if this is something related to the code change or related to the unit test (this is the first cpp unit test I am writing so..)

@roystgnr Would you mind taking a quick look and letting me know where I made mistakes that is obvious to you? Thanks!!

**Comment by roystgnr** [comment]
> I tried numerous times adding a unit test for this code change and it would not work.

That's good!  The best unit tests are the ones that prevent regressions of bugs that have already happened.

I can replicate this; looking at it today.

**Comment by roystgnr** [comment]
Oh, or maybe I should have refreshed the page first.  Looks like CI thinks you already fixed it.  Reviewing today, then.

**Comment by dewenyushu** [comment]
> 

Great! Really appreciate it!! - I am not familiar with unit tests so please bare with me if there are some stupid mistakes

**Comment by dewenyushu** [comment]
> > I tried numerous times adding a unit test for this code change and it would not work.
> 
> That's good! The best unit tests are the ones that prevent regressions of bugs that have already happened.
> 
> I can replicate this; looking at it today.

Great! Really appreciate it!! - I am not familiar with unit tests so please bare with me if there are some stupid mistakes

File: tests/mesh/boundary_info.C [hunk 3]
```diff
@@ -474,6 +478,117 @@ public:
   }
 #endif // LIBMESH_ENABLE_DIRICHLET
 
+#ifdef LIBMESH_ENABLE_AMR
+  void testBoundaryOnChildrenElements()
+  {
+    LOG_UNIT_TEST;
+
+    std::unique_ptr<Mesh> mesh;
```
Comment (roystgnr) [inline tests/mesh/boundary_info.C:486]: Just put the Mesh on the stack here?

But that's a quibble; this looks like an excellent first unit test.  Not yet *sufficient*, given the failure cases above, but a good start.

**Review by roystgnr — CHANGES_REQUESTED** [review]
We're getting closer, at least?

**Comment by dewenyushu** [comment]
close this one due to #3477

**Pull request closed without merging**